### PR TITLE
windowing/gbm: cleanup and OO structure

### PR DESCRIPTION
**Pull request by lrusak**
some streamlining and moving away from static functions.

@Kwiboo for review

@a1rwulf you can also have a look over please as you asked about this before :)

I also have the atomic work done. I'd like to get this in and then get atomic in.

### CODE REVIEW

File: xbmc/windowing/gbm/WinSystemGbm.h [hunk 1]
```diff
@@ -60,11 +60,12 @@ class CWinSystemGbm : public CWinSystemBase
   virtual void Register(IDispResource *resource);
   virtual void Unregister(IDispResource *resource);
 
+  drm m_drm;
+
 protected:
-  CDRM m_DRM;
+  CDRMUtils *m_DRM;
```
Comment (garbear) [inline xbmc/windowing/gbm/WinSystemGbm.h]: this can be `std::unique_ptr<CDRMUtils> m_DRM`, then in the class you do:

```c++
m_DRM.reset(new CDRMLegacy(&m_drm, &m_gbm));
if (!m_DRM->InitDrm())
{
  m_DRM.reset();
  ...
}
```

Comment (garbear) [reply xbmc/windowing/gbm/WinSystemGbm.h]: Or even better, initialize m_DRM in the constructor to provide the assurance that it's never a nullptr

Comment (lrusak) [reply xbmc/windowing/gbm/WinSystemGbm.h]: Thanks @garbear I'll do what you say in your first comment because when atomic support is added we need to fallback to legacy if atomic fails.

File: xbmc/windowing/gbm/DRMUtils.cpp [hunk 2]
```diff
@@ -35,14 +35,12 @@
...
-
 static drmModeResPtr m_drm_resources = nullptr;
 static drmModeCrtcPtr m_orig_crtc = nullptr;
 
-struct drm *CDRMUtils::GetDrm()
+CDRMUtils *CDRMUtils::GetDrm()
 {
-  return m_drm;
+  return this;
 }
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/DRMUtils.cpp]: IMHO this is what you most likely not want to do.
I don't see any use of this getter, so I'd suggest to remove it.

File: xbmc/windowing/gbm/GBMUtils.cpp [hunk 3]
```diff
@@ -21,18 +21,48 @@
 #include "GBMUtils.h"
 #include "utils/log.h"
 
-bool CGBMUtils::InitGbm(struct gbm *gbm, int hdisplay, int vdisplay)
+bool CGBMUtils::CreateDevice(struct gbm *gbm, int fd)
 {
-  gbm->width = hdisplay;
-  gbm->height = vdisplay;
+  if (gbm->device)
+    CLog::Log(LOGWARNING, "CGBMUtils::%s - device already created", __FUNCTION__);
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/GBMUtils.cpp]: We should early return in this case.
If early return is wrong, we should at least `gbm_device_destroy(gbm->device);` before we call create again, otherwise this could leak memory.

File: xbmc/windowing/gbm/GBMUtils.cpp [hunk 4]
```diff
@@ -21,18 +21,48 @@
...
+    return false;
+  }
+
+  return true;
+}
+
+void CGBMUtils::DestroyDevice(struct gbm *gbm)
+{
+  if (!gbm->device)
+    CLog::Log(LOGWARNING, "CGBMUtils::%s - device already destroyed", __FUNCTION__);
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/GBMUtils.cpp]: I don't think that you need to log this case.

File: xbmc/windowing/gbm/GBMUtils.cpp [hunk 5]
```diff
@@ -21,18 +21,48 @@
...
+  {
+    gbm_device_destroy(gbm->device);
+    gbm->device = nullptr;
+  }
+}
+
+bool CGBMUtils::CreateSurface(struct gbm *gbm, int width, int height)
+{
+  if (gbm->surface)
+    CLog::Log(LOGWARNING, "CGBMUtils::%s - surface already created", __FUNCTION__);
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/GBMUtils.cpp]: Same here, not sure what's the right thing to do, but this could result in leaks.

File: xbmc/windowing/gbm/GBMUtils.cpp [hunk 6]
```diff
@@ -45,16 +75,36 @@ bool CGBMUtils::InitGbm(struct gbm *gbm, int hdisplay, int vdisplay)
...
 }
 
-void CGBMUtils::DestroyGbm(struct gbm *gbm)
+void CGBMUtils::DestroySurface(struct gbm *gbm)
 {
-  if(gbm->surface)
+  if (!gbm->surface)
+    CLog::Log(LOGWARNING, "CGBMUtils::%s - surface already destroyed", __FUNCTION__);
+
+  if (gbm->surface)
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/GBMUtils.cpp]: You should also check if gbm itself is not a nullptr.

Comment (a1rwulf) [reply xbmc/windowing/gbm/GBMUtils.cpp]: There's a lot more spots that check gbm->someptr, the question is if you prefer to hard fail in all those cases or add if (!gbm) checks everywhere.

File: xbmc/windowing/gbm/DRMUtils.cpp (hunk 2 above)
Comment (lrusak) [reply xbmc/windowing/gbm/DRMUtils.cpp]: leftovers, will remove.

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 3 above)
Comment (lrusak) [reply xbmc/windowing/gbm/GBMUtils.cpp]: @Kwiboo did you run into a case where gbm->device was already created?

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 4 above)
Comment (lrusak) [reply xbmc/windowing/gbm/GBMUtils.cpp]: I agree, will remove

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 5 above)
Comment (lrusak) [reply xbmc/windowing/gbm/GBMUtils.cpp]: will remove

Comment (a1rwulf) [reply xbmc/windowing/gbm/GBMUtils.cpp]: Sorry, this is misleading the "Same here" applies to the comment that's more above.
I don't know if it can happen that the surface is already created and you enter this again, but wether we should early return or destroy the surface before it is created a second time.

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 3 above)
Comment (Kwiboo) [reply xbmc/windowing/gbm/GBMUtils.cpp]: Not really, but it looked like create surface or lock front buffer could have been called without destroy surface or release buffer, so added logging for all bad gbm calls in my WIP commit to help spot any issues :)

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 6 above)
Comment (Kwiboo) [reply xbmc/windowing/gbm/GBMUtils.cpp]: Passing `gbm` around should not be needed in the first place.
Original intent was to merge the `struct gbm` with `CGBMUtils` and convert all static methods to member methods, but did not get that far.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp [hunk 7]
```diff
@@ -23,7 +23,8 @@
 #include "cores/VideoPlayer/VideoRenderers/RenderCapture.h"
 #include "cores/VideoPlayer/VideoRenderers/RenderFactory.h"
 #include "utils/log.h"
-#include "windowing/gbm/DRMUtils.h"
+
+CWinSystemGbmGLESContext *CRendererDRMPRIME::m_pWinSystem = nullptr;
```
Comment (Kwiboo) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp]: We can init this in header file like the other member variables.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.h [hunk 8]
```diff
@@ -24,6 +24,7 @@
 #include "cores/VideoPlayer/DVDStreamInfo.h"
 #include "cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodec.h"
 #include "cores/VideoPlayer/Process/VideoBuffer.h"
+#include "windowing/gbm/WinSystemGbmGLESContext.h"
```
Comment (Kwiboo) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.h]: It does not look like this include is needed in this header file.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.h [hunk 9]
```diff
@@ -63,6 +64,8 @@ class CRendererDRMPRIME
   int m_iLastRenderBuffer = -1;
   static const int m_numRenderBuffers = 4;
 
+  static CWinSystemGbmGLESContext *m_pWinSystem;
+
```
Comment (Kwiboo) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.h]: `m_pWinSystem = nullptr`

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 4 above)
Comment (Kwiboo) [reply xbmc/windowing/gbm/GBMUtils.cpp]: Please keep logging, helps spot issues until gbm windowing is fully stable

File: xbmc/windowing/gbm/GBMUtils.cpp (hunk 5 above)
Comment (Kwiboo) [reply xbmc/windowing/gbm/GBMUtils.cpp]: Please keep logging for these kind of bad behaviors, a call to this method in this state should never happen, early return or a destroy call without logging will only hide a real issue.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.h (hunk 9 above)
Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.h]:
```
In file included from /home/lukas/libreelec/build.LibreELEC-iMX6.arm-9.0-devel/kodi-02c0008/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp:22:0:
../xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.h:67:36: error: 'constexpr' needed for in-class initialization of static data member 'CWinSystemGbmGLESContext* CRendererDRMPRIME::m_pWinSystem' of non-integral type [-fpermissive]
   static CWinSystemGbmGLESContext *m_pWinSystem = nullptr;
```

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp (hunk 7 above)
Comment (fritsch) [reply xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIME.cpp]: No we can't :-)

**Comment by lrusak** [comment]
moved CGBMUtils to OO structure as well. Please have a look @Kwiboo

**Comment by Kwiboo** [comment]
Looks good after a quick review on mobile, will test and look closer later tonight.

There is an issue with Tinker Board where the last overlay plane is selected and not the first in current code, do you want a separate PR for a fix or include in this?

File: xbmc/windowing/gbm/WinSystemGbm.cpp [hunk 10]
```diff
@@ -40,23 +43,35 @@ CWinSystemGbm::CWinSystemGbm() :
...
-    CLog::Log(LOGERROR, "CWinSystemGbm::%s - failed to initialize DRM", __FUNCTION__);
+    CLog::Log(LOGERROR, "CWinSystemGbm::%s - failed to initialize Legacy DRM", __FUNCTION__);
+    m_DRM.reset();
     return false;
   }
 
-  m_nativeDisplay = m_gbm.dev;
+  if (!m_GBM->CreateDevice(m_DRM->m_fd))
+  {
+    m_GBM.reset();
```
Comment (Kwiboo) [inline xbmc/windowing/gbm/WinSystemGbm.cpp]: `m_GBM` is never reassigned, could cause issues if `InitWindowSystem` is called a second time by Kodi

File: xbmc/windowing/gbm/WinSystemGbm.cpp [hunk 11]
```diff
@@ -65,27 +80,29 @@ bool CWinSystemGbm::CreateNewWindow(const std::string& name,
                                     bool fullScreen,
                                     RESOLUTION_INFO& res)
 {
-  if(!CDRMUtils::SetMode(res))
+
+
```
Comment (Kwiboo) [inline xbmc/windowing/gbm/WinSystemGbm.cpp]: unnecessary whitespace added

File: xbmc/windowing/gbm/WinSystemGbm.cpp [hunk 12]
```diff
@@ -138,23 +155,36 @@ bool CWinSystemGbm::ResizeWindow(int newWidth, int newHeight, int newLeft, int n
...
+  if(!m_DRM->SetMode(res))
   {
     CLog::Log(LOGERROR, "CWinSystemGbm::%s - failed to set DRM mode", __FUNCTION__);
     return false;
   }
 
-  return m_DRM.SetVideoMode(res);
+  struct gbm_bo *bo = m_GBM->LockFrontBuffer();
+
+  if(!m_DRM->SetVideoMode(res, bo))
```
Comment (Kwiboo) [inline xbmc/windowing/gbm/WinSystemGbm.cpp]: `ReleaseBuffer` is never called if `SetVideoMode` fails, recommend changing to something like:
```
struct gbm_bo *bo = m_GBM->LockFrontBuffer();
bool result = m_DRM->SetVideoMode(res, bo);
m_GBM->ReleaseBuffer();
return result;
```

File: xbmc/windowing/gbm/WinSystemGbm.cpp (hunk 10 above)
Comment (lrusak) [reply xbmc/windowing/gbm/WinSystemGbm.cpp]: I don't really understand. I don't think that can ever happen (not now at least).
